lep/DRAFT/011_Performance_Monitor_Goals_and_Metrics.md

```diff
@@ -0,0 +1,85 @@
+#LEP 011 - Performance Monitor Goals and Metrics
+
+Date:   March 24, 2015
+
+Status: Draft
+
+Author: fffw
+
+## Abstract
+
+To continuously monitor the performance of Lantern network, we need to define dimensions and metrics clearly, and collect data to fulfill the goal at both given and get mode.
+
+We must also keep user's anonymity.
+
+## Proposal
+
+The output, is a multi-dimensional model which can pivot from any dimension. The presentation solution is TBD, hopefully there are existing services or open source projects.
```

Comment:
Note - statshub is already able to roll up data to multiple dimensions, however it only allows pivoting on one dimension at a time.  A proper warehousing solution would actually allow pivoting on multiple dimensions at the same time (for example geography plus time of day).  I was intending to make such enhancements to statshub but never got around to it.

BigQuery already has the ability to pivot on multiple dimensions (GROUP BY dima, dimb, etc) however our current data model doesn't facilitate that.  Statshub was originally designed to be reactive and be able to feed realtime updates to the UI, which is part of why we ended up with Redis and the model that we have.  If we are willing to remove the requirement for realtime updates, we can dramatically simplify statshub and basically just use it as a data collector to feed the data into a [star schema](http://en.wikipedia.org/wiki/Star_schema) stored in BigQuery.

Some background on BigQuery - BigQuery is an MPP column-oriented database that is fast enough to do multi-dimensional rollups on raw fact tables.  This is different from traditional OLAP systems that would attempt to precompute rollups.  Whatever we end up doing, I strongly recommend sticking with BigQuery for the data warehouse because it works extremely well in that capacity.